core/src/main/java/com/graphhopper/routing/ch/EdgeBasedNodeContractor.java

```diff
@@ -0,0 +1,550 @@
+        private Stats stats = new Stats();
+
+        @Override
+        public void handleShortcut(CHEntry edgeFrom, CHEntry edgeTo) {
+            addShortcut(edgeFrom, edgeTo);
+        }
+
+        @Override
+        public Stats getStats() {
+            return stats;
+        }
+
+        @Override
+        public String getAction() {
+            return "add";
+        }
+    }
+
+    private class CountingShortcutHandler implements ShortcutHandler {
+        private Stats stats = new Stats();
+
+        @Override
+        public void handleShortcut(CHEntry edgeFrom, CHEntry edgeTo) {
+            int fromNode = edgeFrom.parent.adjNode;
+            int toNode = edgeTo.adjNode;
+            int firstOrigEdge = edgeFrom.getParent().incEdge;
+            int lastOrigEdge = edgeTo.incEdge;
+
+            // check if this shortcut already exists
+            final CHEdgeIterator iter = existingShortcutExplorer.setBaseNode(fromNode);
+            while (iter.next()) {
+                if (isSameShortcut(iter, toNode, firstOrigEdge, lastOrigEdge)) {
+                    // this shortcut exists already, maybe its weight will be updated but we should not count it as
+                    // a new edge
+                    return;
+                }
+            }
+
+            // this shortcut is new --> increase counts
+            numShortcuts++;
+            numOrigEdges += getOrigEdgeCount(edgeFrom.edge) + getOrigEdgeCount(edgeTo.edge);
+        }
+
+        @Override
+        public Stats getStats() {
+            return stats;
+        }
+
+        @Override
+        public String getAction() {
+            return "count";
+        }
+    }
+
+    public static class Params {
+        // todo: optimize
+        private float edgeQuotientWeight = 1;
+        private float originalEdgeQuotientWeight = 3;
+        private float hierarchyDepthWeight = 2;
+    }
+
+    private static class Stats {
+        int nodes;
+        long loopsAvoided;
+        StopWatch stopWatch = new StopWatch();
+
+        @Override
+        public String toString() {
+            return String.format("time: %7.2fs, nodes-handled: %10s, loopsAvoided: %10s",
+                    stopWatch.getCurrentSeconds(), nf(nodes), nf(loopsAvoided));
+        }
+    }
+
+    private interface SearchStrategy {
+        void findAndHandleShortcuts(int node);
+
+        String getStatisticsString();
+
+        void resetStats();
+
+    }
+
+    private class AggressiveStrategy implements SearchStrategy {
+        @Override
+        public String getStatisticsString() {
+            return witnessPathSearcher.getStatisticsString();
+        }
+
+        @Override
+        public void resetStats() {
+            witnessPathSearcher.resetStats();
+        }
+
+        @Override
+        public void findAndHandleShortcuts(int node) {
+            LOGGER.debug("Finding shortcuts (aggressive) for node {}, required shortcuts will be {}ed", node, activeShortcutHandler.getAction());
```

Comment:
Using debug and trace did not really make sense here, because both logs would very frequent when they are enabled. I used this mostly for debugging. I set all to trace now.